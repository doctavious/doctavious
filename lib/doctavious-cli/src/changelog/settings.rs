use std::str::FromStr;

use lazy_static::lazy_static;
use regex::Regex;
use strum::{Display, EnumIter, EnumString, VariantNames};

lazy_static! {
    static ref CHANGELOG_RANGE_REGEX: Regex = Regex::new(r"^.+\.\..+$").unwrap();
}

#[derive(Clone, Debug, Display, EnumIter, VariantNames, PartialEq)]
pub enum ChangelogRange {
    Current,
    Latest,
    Unreleased,
    Range(String),
}

impl FromStr for ChangelogRange {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "current" => Ok(ChangelogRange::Current),
            "latest" => Ok(ChangelogRange::Latest),
            "unreleased" => Ok(ChangelogRange::Unreleased),
            _ => {
                if CHANGELOG_RANGE_REGEX.is_match(s) {
                    Ok(ChangelogRange::Range(s.to_string()))
                } else {
                    Err(
                        "Invalid changelog range. Value should be current, latest, unreleased, or \
                        in the format of <START>..<END>"
                            .to_string(),
                    )
                }
            }
        }
    }
}

#[derive(Clone, Copy, Debug, Default, Display, EnumIter, EnumString, VariantNames, PartialEq)]
pub enum ChangelogCommitSort {
    Newest,
    #[default]
    Oldest,
}

impl ChangelogCommitSort {
    #[must_use]
    pub const fn variants() -> &'static [&'static str] {
        <Self as strum::VariantNames>::VARIANTS
    }
}
