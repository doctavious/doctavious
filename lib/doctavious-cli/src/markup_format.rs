use std::collections::HashMap;
use std::path::Path;
use std::str::FromStr;

use lazy_static::lazy_static;
use serde::{Deserialize, Deserializer, Serialize, Serializer};
use strum::{Display, EnumIter, EnumString, IntoEnumIterator, VariantNames};

use crate::{CliResult, DoctaviousCliError};

lazy_static! {
    pub static ref MARKUP_FORMAT_EXTENSIONS: HashMap<&'static str, MarkupFormat> = {
        let mut map = HashMap::new();
        for markup_format in MarkupFormat::iter() {
            map.insert(markup_format.extension(), markup_format.to_owned());
        }
        map
    };
}

#[derive(Clone, Copy, Debug, Display, EnumIter, EnumString, VariantNames, PartialEq)]
#[non_exhaustive]
pub enum MarkupFormat {
    #[strum(serialize = "adoc")]
    Asciidoc,
    #[strum(serialize = "md")]
    Markdown,
    // TODO: Other(str)?
}

impl MarkupFormat {
    pub(crate) fn extension(&self) -> &'static str {
        return match self {
            Self::Asciidoc => "adoc",
            Self::Markdown => "md",
        };
    }

    pub(crate) fn leading_header_character(&self) -> char {
        return match self {
            Self::Asciidoc => '=',
            Self::Markdown => '#',
        };
    }

    #[must_use]
    pub const fn variants() -> &'static [&'static str] {
        <Self as strum::VariantNames>::VARIANTS
    }

    pub fn from_path<P: AsRef<Path>>(path: P) -> CliResult<Self> {
        let path = path.as_ref();
        // has to be a better way to do this
        let extension = path
            .extension()
            .ok_or(DoctaviousCliError::UnknownMarkupExtension(
                path.to_string_lossy().to_string(),
            ))?
            .to_str()
            .ok_or(DoctaviousCliError::UnknownMarkupExtension(
                path.to_string_lossy().to_string(),
            ))?;
        Ok(Self::from_str(extension)?)
    }
}

impl Default for MarkupFormat {
    fn default() -> Self {
        Self::Markdown
    }
}

impl Serialize for MarkupFormat {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let s = self.extension();
        s.serialize(serializer)
    }
}

impl<'de> Deserialize<'de> for MarkupFormat {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        let s = String::deserialize(deserializer)?;
        let markup_format = <MarkupFormat as FromStr>::from_str(&s).unwrap_or_else(|e| {
            eprintln!(
                "Error when parsing {}, fallback to default settings. Error: {:?}\n",
                s, e
            );
            MarkupFormat::default()
        });
        Ok(markup_format)
    }
}
