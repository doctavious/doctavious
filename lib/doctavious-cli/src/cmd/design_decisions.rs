use std::fmt::{Display, Formatter, Pointer};
use std::fs;
use std::io::ErrorKind;
use std::path::{Path, PathBuf};
use std::str::FromStr;

use markup::{MARKUP_FORMAT_EXTENSIONS, MarkupFormat};
use scm::drivers::{Scm, ScmRepository};
use thiserror::Error;
use unidecode::unidecode;
use walkdir::{DirEntry, WalkDir};

use crate::errors::{CliResult, DoctaviousCliError};
use crate::file_structure::FileStructure;
use crate::settings::DEFAULT_TEMPLATE_DIR;

pub mod adr;
pub mod rfd;

#[remain::sorted]
#[derive(Debug, Error)]
pub enum DesignDecisionErrors {
    #[error("design doc directory already initialized")]
    DesignDocAlreadyInitialized,

    #[error("design doc directory already exists")]
    DesignDocDirectoryAlreadyExists,

    #[error("invalid design doc directory. Should be utf-8")]
    DesignDocDirectoryInvalid,

    #[error("invalid link reference")]
    InvalidLinkReference,

    /// Error that may occur while reserving ADR/RFD number.
    #[error("{0} has already been reserved")]
    ReservedNumberError(u32),

    #[error("Unknown design document: {0}")]
    UnknownDesignDocument(String),
}

#[remain::sorted]
#[derive(Debug, Clone)]
pub enum LinkReference {
    FileName(String),
    Number(u32),
    Path(PathBuf),
}

impl FromStr for LinkReference {
    type Err = DoctaviousCliError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        Ok(if let Ok(num) = s.parse::<u32>() {
            Self::Number(num)
        } else {
            if let Ok(path) = PathBuf::from_str(s) {
                if path.is_file() {
                    return Ok(Self::Path(path));
                }
            }
            Self::FileName(s.to_string())
        })
    }
}

impl LinkReference {
    pub fn get_record(&self, cwd: &Path) -> Option<PathBuf> {
        let reference = match self {
            Self::FileName(file) => file.to_string(),
            Self::Number(num) => format_number(num),
            Self::Path(path) => return Some(path.to_owned()),
        };

        get_records(cwd)
            .filter_map(|e| {
                if e.path().to_string_lossy().contains(&reference) {
                    Some(e.path().to_path_buf())
                } else {
                    None
                }
            })
            .next()
    }
}

impl Display for LinkReference {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let s = match self {
            Self::FileName(f) => f.to_string(),
            Self::Number(n) => n.to_string(),
            Self::Path(p) => p.to_string_lossy().to_string(),
        };

        write!(f, "{s}")
    }
}

pub(crate) fn list(dir: &Path, format: MarkupFormat) -> CliResult<Vec<PathBuf>> {
    let mut paths: Vec<_> = get_records(dir)
        .filter(|e| {
            if let Some(extension) = e.path().extension() {
                return extension.to_string_lossy() == format.extension();
            }
            false
        })
        .map(|e| e.path().to_path_buf())
        .collect();

    paths.sort_by(|a, b| a.file_name().cmp(&b.file_name()));
    Ok(paths)
}

fn get_records(cwd: &Path) -> impl Iterator<Item = DirEntry> {
    WalkDir::new(cwd)
        .into_iter()
        .filter_entry(|e| {
            if e.path().is_dir() {
                return e.file_name().to_string_lossy() != DEFAULT_TEMPLATE_DIR;
            }

            true
        })
        .filter_map(Result::ok)
        .filter(|e| e.file_type().is_file())
        .filter(|e| is_valid_file(&e.path()))
}

pub(crate) fn format_number(number: &u32) -> String {
    format!("{:0>4}", number)
}

// TODO: is there a more concise way to do this?
pub(crate) fn build_path(
    dir: &Path,
    title: &str,
    reserved_number: &str,
    extension: MarkupFormat,
    file_structure: FileStructure,
) -> PathBuf {
    match file_structure {
        FileStructure::Flat => {
            let slug = slugify(&title);
            let file_name = format!("{}-{}", reserved_number, slug);
            Path::new(dir)
                .join(file_name)
                .with_extension(extension.to_string())
        }

        FileStructure::Nested => Path::new(dir)
            .join(&reserved_number)
            .join("README")
            .with_extension(extension.to_string()),
    }
}

pub(crate) fn reserve_number(
    dir: &Path,
    number: Option<u32>,
    file_structure: FileStructure,
) -> CliResult<u32> {
    if let Some(i) = number {
        if is_number_reserved(dir, i, file_structure) {
            // TODO: the prompt to overwrite be here?
            eprintln!("{} has already been reserved in directory {:?}", i, dir);
            return Err(DoctaviousCliError::DesignDecisionErrors(
                DesignDecisionErrors::ReservedNumberError(i),
            ));
        }
        Ok(i)
    } else {
        Ok(get_next_number(dir, file_structure))
    }
}

pub(crate) fn is_number_reserved(dir: &Path, number: u32, file_structure: FileStructure) -> bool {
    get_allocated_numbers(dir, file_structure).contains(&number)
}

pub(crate) fn get_allocated_numbers(dir: &Path, file_structure: FileStructure) -> Vec<u32> {
    match file_structure {
        FileStructure::Flat => get_allocated_numbers_via_flat_files(dir),
        FileStructure::Nested => get_allocated_numbers_via_nested(dir),
    }
}

// TODO: do we want a ReservedNumber type?
// TODO: would be nice to do this via an Iterator but having trouble with empty
// expected struct `std::iter::Map`, found struct `std::iter::Empty`
// using vec for now
pub(crate) fn get_allocated_numbers_via_nested(dir: &Path) -> Vec<u32> {
    match fs::read_dir(dir) {
        Ok(files) => {
            return files
                .filter_map(Result::ok)
                .filter_map(|e| {
                    if e.path().is_dir() {
                        e.file_name().to_string_lossy().parse::<u32>().ok()
                    } else {
                        None
                    }
                })
                .collect();
        }
        Err(e) if e.kind() == ErrorKind::NotFound => {
            return Vec::new();
        }
        Err(e) => {
            // TODO: dont panic here
            panic!("Error reading directory {:?}. Error: {}", dir, e);
        }
    }
}

// TODO: extract this to list or a list helper
// TODO: would be nice to do this via an Iterator but having trouble with empty
// expected struct `std::iter::Map`, found struct `std::iter::Empty`
// using vec for now
pub(crate) fn get_allocated_numbers_via_flat_files(dir: &Path) -> Vec<u32> {
    let mut allocated_numbers = Vec::new();
    for entry in WalkDir::new(&dir)
        .into_iter()
        .filter_entry(|e| {
            if e.path().is_dir() {
                return e.file_name().to_string_lossy() != DEFAULT_TEMPLATE_DIR;
            }

            true
        })
        .filter_map(Result::ok)
        .filter(|e| e.file_type().is_file())
        .filter(|e| is_valid_file(&e.path()))
    {
        // The only way I can get this to pass the borrow checker is first mapping
        // to file_name and then doing the rest.
        // I'm probably doing this wrong and should review later
        let file_name = entry.file_name();
        let ss = file_name.to_str().unwrap();
        let first_space_index = ss.find("-").expect("didnt find a hyphen");
        let num: String = ss.chars().take(first_space_index).collect();
        allocated_numbers.push(num.parse::<u32>().unwrap());
    }

    allocated_numbers
}

pub(crate) fn get_next_number(dir: &Path, file_structure: FileStructure) -> u32 {
    // TODO: revisit iterator
    // get_allocated_numbers(dir, file_structure)
    //     .iter()
    //     .max()
    //     .unwrap_or_default() + 1

    if let Some(max) = get_allocated_numbers(dir, file_structure).iter().max() {
        max + 1
    } else {
        1
    }
}

fn can_reserve(repo: &Scm, number: u32) -> CliResult<bool> {
    if repo.is_dirty()? {
        // TODO: return error
    }

    match repo {
        Scm::Git(_) => {
            if repo.branch_exists(number.to_string().as_str()).is_err() {
                return Err(
                    git2::Error::from_str("branch already exists in remote. Please pull.").into(),
                );
            }
        }
        _ => unimplemented!(),
    }

    Ok(true)
}

pub(crate) fn slugify(string: &str) -> String {
    let separator_char = '-';
    let separator = separator_char.to_string();

    let string: String = unidecode(string.into())
        .to_lowercase()
        .trim_matches(separator_char)
        .replace(' ', &separator);

    let mut slug = Vec::with_capacity(string.len());
    let mut is_sep = true;

    for x in string.chars() {
        match x {
            'a'..='z' | '0'..='9' => {
                is_sep = false;
                slug.push(x as u8);
            }
            _ => {
                if !is_sep {
                    is_sep = true;
                    slug.push(separator_char as u8);
                }
            }
        }
    }

    if slug.last() == Some(&(separator_char as u8)) {
        slug.pop();
    }

    let s = String::from_utf8(slug).unwrap();
    s.trim_end_matches(separator_char).to_string()
}

// TODO: where does this belong
pub(crate) fn is_valid_file(path: &Path) -> bool {
    MARKUP_FORMAT_EXTENSIONS.contains_key(&path.extension().unwrap().to_str().unwrap())
}

#[cfg(test)]
mod tests {}
