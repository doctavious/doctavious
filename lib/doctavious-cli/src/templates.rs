use std::fs;
use std::io::BufRead;
use std::path::{Path, PathBuf};

use crate::markup_format::MarkupFormat;
use crate::templating::TemplateType;


pub(crate) fn get_template(dir: &Path, template_type: TemplateType, extension: &str) -> PathBuf {
    // see if direction defines a custom template
    let custom_template = dir
        .join(template_type.get_default_path())
        .with_extension(extension);

    if custom_template.is_file() {
        custom_template
    } else {
        PathBuf::from(env!("CARGO_MANIFEST_DIR")).join(template_type.get_default_path().with_extension(extension))
    }
}

pub(crate) fn get_template_content(
    dir: &Path,
    template_type: TemplateType,
    extension: &str,
) -> String {
    let template_path = get_template(dir, template_type, extension);
    // TODO: we shouldnt panic here
    fs::read_to_string(&template_path).expect(&format!(
        "failed to read file {}.",
        &template_path.to_string_lossy()
    ))
}

pub(crate) fn get_title<R>(rdr: R, markup_format: MarkupFormat) -> String
where
    R: BufRead,
{
    // TODO: swap this implementation for AST when ready
    let leading_char = markup_format.leading_header_character();
    for line in rdr.lines() {
        let line = line.unwrap();
        if line.starts_with(leading_char) {
            let last_hash = line
                .char_indices()
                .skip_while(|&(_, c)| c == leading_char)
                .next()
                .map_or(0, |(idx, _)| idx);

            // Trim the leading hashes and any whitespace
            return line[last_hash..].trim().to_string();
        }
    }

    String::new()
}
