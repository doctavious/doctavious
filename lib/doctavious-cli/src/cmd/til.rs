use std::collections::BTreeMap;
use std::fs::File;
use std::io::{BufReader, ErrorKind};
use std::path::{Path, PathBuf};
use std::str::FromStr;
use std::{env, fs};

use chrono::{DateTime, Local};
use directories::UserDirs;
use serde::Serialize;
use tracing::debug;
use walkdir::{DirEntry, WalkDir};

use crate::cmd::design_decisions::is_valid_file;
use crate::files::{ensure_path, friendly_filename};
use crate::markup_format::MarkupFormat;
use crate::settings::{Config, SettingErrors, TilSettings, DEFAULT_TIL_DIR};
use crate::templates::{get_template, get_template_content, get_title};
use crate::templating::{TemplateContext, TemplateType, Templates, TilTemplateType};
use crate::{edit, CliResult, DoctaviousCliError};

#[derive(Clone, Debug, Serialize)]
struct TilEntry {
    topic: String,
    title: String,
    file_name: String,
    description: String,
    date: String,
}

// TODO: should probably be part of CLI docs.
/// Init TIL settings using the following rules
///
/// If `local` is true create settings in either `cwd` or the current directory
/// If `local` if false create global settings with `til.dir` as either `cwd` or default til directory
///
/// Global settings exist at the following locations
/// - on Linux: /home/<user>/.config/com.doctavious.cli/doctavious.toml
/// - on Windows: C:\Users\<user>\AppData\Roaming\com.doctavious.cli\doctavious.toml
/// - on macOS: /Users/<user>/Library/Application Support/com.doctavious.cli/doctavious.toml
pub fn init(cwd: Option<&Path>, format: MarkupFormat, local: bool) -> CliResult<PathBuf> {
    let cwd = cwd.and_then(|p| Some(p.to_path_buf()));
    let (mut config, til_dir) = if local {
        let path = cwd.unwrap_or(env::current_dir()?);
        (Config::from_path_or_default(&path), path)
    } else {
        let til_dir = cwd.unwrap_or(
            UserDirs::new()
                .expect("Could not get user directory")
                .home_dir()
                .join(DEFAULT_TIL_DIR),
        );
        (Config::get_global_or_default(), til_dir)
    };

    if config.settings.til_settings.is_some() {
        return Err(DoctaviousCliError::SettingError(
            SettingErrors::AlreadyInitialized("til".to_string()),
        ));
    }

    create_til_dir(&til_dir)?;

    config.settings.til_settings = Some(TilSettings {
        // dont need to include dir in a local config
        dir: if local {
            None
        } else {
            Some(til_dir.to_owned())
        },
        template_format: format,
    });

    config.save()?;

    Ok(til_dir)
}

fn create_til_dir(path: &Path) -> CliResult<()> {
    match fs::create_dir_all(path) {
        Ok(_) => Ok(()),
        Err(e) if e.kind() == ErrorKind::AlreadyExists => {
            debug!("the directory {} already exists", path.to_string_lossy());
            Ok(())
        }
        Err(e) => {
            debug!(
                "Error occurred creating directory {}: {}",
                path.to_string_lossy(),
                e
            );
            Err(e.into())
        }
    }
}

// If `cwd` is provided but it doesnt have settings should we check if global settings exist?
// Should we allow people to customize where ToC write to?
/// Create new TIL post
///
/// If `cwd` is provided a create new TIL in that directory. Use settings from `cwd` if present, otherwise fallback to
/// global if present and default when not.
/// If `cwd` is not present see if current directory has settings, if present use, otherwise see if global
/// settings are present, if they are use them, otherwise default to current directory.
pub fn new(
    cwd: Option<&Path>,
    post: String,
    tags: Option<Vec<String>>,
    toc: bool,
) -> CliResult<PathBuf> {
    let config = get_config(cwd.as_deref())?;

    let Some((category, title)) = post.split_once("/") else {
        todo!()
    };

    let til_dir = if let Some(til_dir) = config.settings.get_til_dir() {
        til_dir
    } else {
        if let Some(cwd) = cwd {
            cwd.to_path_buf()
        } else {
            env::current_dir().expect("Unable to get current directory")
        }
    };

    let format = MarkupFormat::from_path(Path::new(title))
        .ok()
        .unwrap_or(config.settings.get_til_template_format(None));

    let path = Path::new(&til_dir)
        .join(category)
        .join(friendly_filename(title))
        .with_extension(format.extension());

    if path.is_file() {
        return Err(DoctaviousCliError::TilAlreadyExists);
    }

    ensure_path(&path)?;

    let template = get_template(
        &til_dir,
        TemplateType::Til(TilTemplateType::Post),
        &format.extension(),
    );

    let template_content = fs::read_to_string(&template).expect(&format!(
        "failed to read file {}.",
        &template.to_string_lossy()
    ));

    let mut context = TemplateContext::new();
    context.insert("title", &friendly_title(title));
    context.insert("date", &Local::now().format("%Y-%m-%d").to_string());
    if let Some(tags) = tags {
        context.insert("tags", &tags.join(" ").to_string());
    }

    let rendered = Templates::one_off(template_content.as_str(), context, false)?;

    let edited = edit::edit(&rendered)?;
    fs::write(&path, edited)?;

    if toc {
        generate_toc(&til_dir, format)?;
    }

    Ok(path)
}

pub fn list(cwd: &Path) -> CliResult<Vec<PathBuf>> {
    let paths: Vec<_> = get_posts(cwd, None)
        .map(|e| e.path().to_path_buf())
        .collect();

    Ok(paths)
}

pub fn open(cwd: Option<&Path>, post: String) -> CliResult<PathBuf> {
    let config = get_config(cwd)?;

    // TODO: extract to fn
    let til_dir = if let Some(til_dir) = config.settings.get_til_dir() {
        til_dir
    } else {
        if let Some(cwd) = cwd {
            cwd.to_path_buf()
        } else {
            env::current_dir().expect("Unable to get current directory")
        }
    };

    let Some((topic, title)) = post.split_once("/") else {
        todo!()
    };

    let path = get_posts(&til_dir, Some(topic))
        .filter_map(|e| {
            if e.path().to_string_lossy().contains(title) {
                Some(e.path().to_path_buf())
            } else {
                None
            }
        })
        .next();

    if let Some(post_path) = path {
        let edited = edit::edit_path(&post_path)?;
        fs::write(&post_path, edited)?;
        Ok(post_path)
    } else {
        todo!()
    }
}

pub fn render() -> CliResult<()> {
    Ok(())
}

fn get_config(cwd: Option<&Path>) -> CliResult<Config> {
    Ok(if let Some(cwd) = cwd {
        Config::from_path_or_default(&cwd)
    } else {
        let local_config = Config::from_path_or_default(&env::current_dir()?);
        if local_config.is_default_settings {
            Config::get_global().unwrap_or_else(|_| local_config)
        } else {
            local_config
        }
    })

    // TODO: why does this cause tests to hang?
    // let path = cwd.and_then(|p| Some(p.to_path_buf())).unwrap_or(env::current_dir()?);
    // let local_config = Config::from_path_or_default(&path);
    // Ok(if local_config.is_default_settings {
    //     Config::get_global().unwrap_or_else(|_| local_config)
    // } else {
    //     local_config
    // })
}

fn get_posts(cwd: &Path, topic: Option<&str>) -> impl Iterator<Item = DirEntry> {
    let mut dir = cwd.to_path_buf();
    if let Some(topic) = topic {
        dir = dir.join(topic);
    }

    // TODO: should probably ignore hidden directories but that messes up tests which uses temp_dir
    WalkDir::new(dir)
        .into_iter()
        .filter_map(Result::ok)
        .filter(|e| e.file_type().is_file())
        .filter(|e| is_valid_file(&e.path()))
}

pub fn generate_toc(dir: &Path, format: MarkupFormat) -> CliResult<()> {
    let mut all_tils: BTreeMap<String, Vec<TilEntry>> = BTreeMap::new();
    for entry in get_posts(&dir, None) {
        if let Some(til) = file_to_til_entry(dir, entry) {
            if let Some(tils) = all_tils.get_mut(&til.topic) {
                tils.push(til);
            } else {
                all_tils.insert(til.topic.clone(), vec![til]);
            }
        }
    }

    let mut til_count = 0;
    for topic_tils in all_tils.values() {
        til_count += topic_tils.len();
    }

    let template = get_template_content(
        dir,
        TemplateType::Til(TilTemplateType::ReadMe),
        format.extension(),
    );

    let mut context = TemplateContext::new();
    context.insert("categories_count", &all_tils.keys().len());
    context.insert("til_count", &til_count);
    context.insert("tils", &all_tils);

    let rendered = Templates::one_off(template.as_str(), context, false)?;
    let p = dir.join("README").with_extension(format.extension());
    fs::write(p, &rendered)?;

    Ok(())
}

fn file_to_til_entry(root: &Path, entry: DirEntry) -> Option<TilEntry> {
    let entry_path = entry.path();
    let parent = entry.path().parent()?;
    // skip files that are under til dir
    if root == parent {
        debug!("Skipping {:?}. Not in a topic", entry_path);
        return None;
    }

    let topic = parent.file_name();
    if topic.is_none() {
        debug!("Skipping {:?}. Could not get topic", entry_path);
        return None;
    }

    let title = entry_path.file_stem();
    if title.is_none() {
        debug!("Skipping {:?}. Could not get title", entry_path);
        return None;
    }

    let markup_format = MarkupFormat::from_path(entry_path);
    if markup_format.is_err() {
        debug!("Skipping {:?}. Could not get markup format", entry_path);
        return None;
    }

    let file = File::open(entry_path);
    if file.is_err() {
        debug!("Skipping {:?}. Could not open file", entry_path);
        return None;
    }

    let buffer = BufReader::new(file.ok()?);
    let description = get_title(buffer, markup_format.ok()?);
    let file_name = entry.path().file_name();
    if file_name.is_none() {
        debug!("Skipping {:?}. Could not get file_name", entry_path);
        return None;
    }

    let created = entry.metadata().ok().and_then(|m| m.created().ok());
    if created.is_none() {
        debug!("Skipping {:?}. Could not get created date", entry_path);
        return None;
    }

    Some(TilEntry {
        topic: topic?.to_string_lossy().to_string(),
        title: title?.to_string_lossy().to_string(),
        description,
        file_name: file_name?.to_string_lossy().to_string(),
        date: DateTime::<Local>::from(created?)
            .format("%Y-%m-%d")
            .to_string(),
    })
}

// TODO: where to put this
fn is_hidden(entry: &DirEntry) -> bool {
    entry
        .file_name()
        .to_str()
        .map(|s| s.starts_with("."))
        .unwrap_or(false)
}

fn friendly_title(s: &str) -> String {
    s.split(&['-', '_'])
        .map(|word| {
            let mut chars = word.chars();
            match chars.next() {
                None => String::new(),
                Some(f) => f.to_uppercase().collect::<String>() + chars.as_str(),
            }
        })
        .collect::<Vec<String>>()
        .join(" ")
}


#[cfg(test)]
mod tests {
    use std::fs;
    use std::path::Path;

    use directories::BaseDirs;
    use tempfile::TempDir;

    use crate::cmd::til::{init, list, new, open};
    use crate::files::get_all_files;
    use crate::markup_format::MarkupFormat;
    use crate::settings::Config;

    #[test]
    fn should_successfully_init_local() {
        let dir = TempDir::new().unwrap();
        temp_env::with_vars(
            [("EDITOR", Some(Path::new("./tests/fixtures/noop-editor")))],
            || {
                let path = init(Some(dir.path()), MarkupFormat::Markdown, true)
                    .expect("Should be able to init TIL");

                assert!(path.is_dir());
                let files = get_all_files(dir.path());
                assert_eq!(1, files.len());

                insta::with_settings!({filters => vec![
                    (dir.path().to_str().unwrap(), "[DIR]"),
                ]}, {
                    insta::assert_snapshot!(fs::read_to_string(&files[0]).unwrap());
                });
            },
        );

        dir.close().unwrap();
    }

    // TODO: setup for CI flag
    #[test]
    // #[cfg(ci)]
    fn should_successfully_init_global() {
        let dir = TempDir::new().unwrap();
        temp_env::with_vars(
            [("EDITOR", Some(Path::new("./tests/fixtures/noop-editor")))],
            || {
                let path = init(Some(dir.path()), MarkupFormat::Markdown, false)
                    .expect("Should be able to init TIL");

                assert!(path.is_dir());
                let files = get_all_files(dir.path());

                // should not have any files in the TIL dir after init for global settings
                assert!(files.is_empty());

                let config = Config::get_global().unwrap();
                assert!(!config.is_default_settings);

                let expected_config_path = BaseDirs::new()
                    .unwrap()
                    .config_dir()
                    .join("com.doctavious.cli/doctavious.toml")
                    .to_string_lossy()
                    .to_string();

                assert!(config
                    .path
                    .to_string_lossy()
                    .ends_with(expected_config_path.as_str()));

                insta::with_settings!({filters => vec![
                    (dir.path().to_str().unwrap(), "[DIR]"),
                    (r"\d{4}-\d{2}-\d{2}", "[DATE]")
                ]}, {
                    insta::assert_yaml_snapshot!(&config);
                    insta::assert_snapshot!(fs::read_to_string(&config.path).unwrap());
                });
            },
        );

        dir.close().unwrap();
    }

    #[test]
    fn should_successfully_create_new_post() {
        let dir = TempDir::new().unwrap();
        temp_env::with_vars(
            [("EDITOR", Some(Path::new("./tests/fixtures/noop-editor")))],
            || {
                let path = new(
                    Some(dir.path()),
                    "rust/testing".to_string(),
                    None,
                    false,
                )
                .expect("Should be able to create new post");

                insta::with_settings!({filters => vec![
                    (r"\d{4}-\d{2}-\d{2}", "[DATE]")
                ]}, {
                    insta::assert_snapshot!(fs::read_to_string(path).unwrap());
                });
            },
        );

        dir.close().unwrap();
    }

    #[test]
    fn should_edit_on_new_post() {
        let dir = TempDir::new().unwrap();

        temp_env::with_vars(
            [("EDITOR", Some(Path::new("./tests/fixtures/fake-editor")))],
            || {
                let path = new(
                    Some(dir.path()),
                    "rust/testing".to_string(),
                    None,
                    false,
                )
                .expect("Should be able to create new post");

                let content = fs::read_to_string(&path).unwrap();
                assert!(content.starts_with("EDITOR"));
            },
        );

        dir.close().unwrap();
    }

    #[test]
    fn should_successfully_create_toc() {
        let dir = TempDir::new().unwrap();

        temp_env::with_vars(
            [("EDITOR", Some(Path::new("./tests/fixtures/noop-editor")))],
            || {
                let path = new(
                    Some(dir.path()),
                    "rust/testing".to_string(),
                    None,
                    true,
                )
                .expect("Should be able to create new post and ToC");

                insta::with_settings!({filters => vec![
                    (r"\d{4}-\d{2}-\d{2}", "[DATE]")
                ]}, {
                    insta::assert_snapshot!(fs::read_to_string(dir.path().join("README.md")).unwrap());
                });
            },
        );

        dir.close().unwrap();
    }

    #[test]
    fn should_list() {
        let dir = TempDir::new().unwrap();

        temp_env::with_vars(
            [("EDITOR", Some(Path::new("./tests/fixtures/noop-editor")))],
            || {
                new(
                    Some(dir.path()),
                    "rust/testing".to_string(),
                    None,
                    false,
                )
                .expect("Should be able to create new post");

                new(
                    Some(dir.path()),
                    "baz/foo".to_string(),
                    None,
                    false,
                )
                .expect("Should be able to create new post");

                let all_posts = list(dir.path()).unwrap();

                assert_eq!(2, all_posts.len());
            },
        );

        dir.close().unwrap();
    }

    #[test]
    fn should_open() {
        let dir = TempDir::new().unwrap();

        temp_env::with_vars(
            [("VISUAL", Some(Path::new("./tests/fixtures/fake-visual")))],
            || {
                new(
                    Some(dir.path()),
                    "rust/testing".to_string(),
                    None,
                    false,
                )
                .expect("Should be able to create new post");

                let path = open(Some(dir.path()), "rust/testing".to_string()).unwrap();
                assert!(fs::read_to_string(&path).unwrap().contains("VISUAL"));
            },
        );

        dir.close().unwrap();
    }
}
