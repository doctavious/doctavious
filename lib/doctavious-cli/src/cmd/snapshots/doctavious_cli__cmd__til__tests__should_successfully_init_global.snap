---
source: lib/doctavious-cli/src/cmd/til.rs
expression: "&config"
---
path: /Users/seancarroll/Library/Application Support/com.doctavious.cli/doctavious-test.toml
settings:
  til:
    dir: [DIR]
    template_format: md
is_default_settings: false
